### PR TITLE
Fix #87

### LINKED ISSUE #87: Event delegation issues in v8.1.0
> I just notices you have implemented event delegation in cycle dom. I did not took a close look yet but I noticed it because 8.1.0 broke my app :smile: I used `event.currentTarget` in various places which now gives me a different element since the listener is bound to the drivers root and not to the element itself. Replacing `currentTarget` with `target` made most of my app working again.
>
> But one problem remains: If I use `DOM.select('.myElement').events('click')` and my markup looks like: `div('.myElement', [div('.someChild', 'Hello'), div('.otherChild', 'world')])` the clicks are not recognized if they are targeted at "Hello" or "World" because then event.target is not `.myElement` but `.someChild` thus is not matched by the `.myElement` selector (But it's parent would match it so the event should be accepted). This concerns all listeners on elements which have child elements. I will take a closer look tomorrow and will create a proper issue. 
>
> I think `event.target` is not affected by the event phase. If I am correct for mouse events `target` is always the deepest node in the tree for which `pointer-events` is not `none`. Capture and Bubble phases happen after `event.target` is already determined.
>
> Currently most (probably even all) of my cases could be fixed by setting `pointer-events` to `none` for all my child elements. But I think the correct way of handling it would be to not only check `evt.target` itself but all it's ancestors at https://github.com/cyclejs/cycle-dom/blob/master/src/render-dom.js#L132-L133
>
> I think jQuery even overwrites the event's currentTarget property during it's event dispatching to make the event delegation mostly transparent from the listeners point of view

```js
it('should have currentTarget pointed to the selected parent', function (done) {
    function app() {
      return {
        DOM: Rx.Observable.just(div('.top', [
          h2('.parent', [
            span('.child', 'Hello world')
          ])
        ]))
      };
    }

    const {sinks, sources} = Cycle.run(app, {
      DOM: makeDOMDriver(createRenderTarget())
    });

    sources.DOM.select('.parent').events('click').subscribe(ev => {
      assert.strictEqual(ev.type, 'click');
      assert.strictEqual(ev.target.tagName, 'SPAN');
      assert.strictEqual(ev.target.className, 'child');
      assert.strictEqual(ev.target.textContent, 'Hello world');
      assert.strictEqual(ev.currentTarget.tagName, 'H2');
      assert.strictEqual(ev.currentTarget.className, 'parent');
      sources.dispose();
      done();
    });
    // Make assertions
    sources.DOM.select(':root').observable.skip(1).take(1).subscribe(function (root) {
      const child = root.querySelector('.child');
      assert.notStrictEqual(child, null);
      assert.notStrictEqual(typeof child, 'undefined');
      assert.strictEqual(child.tagName, 'SPAN');
      assert.strictEqual(child.className, 'child');
      assert.doesNotThrow(function () {
        child.click();
      });
    });
  });
```

Plan:
- [x] Override currentTarget so that it passes the test above
- [x] Automatically set useCapture to true for some whitelisted events: focus, blur, mouseenter, etc.
- [x] Change the API from events(eventType, useCapture) to events(eventType, options)

### PR DESCRIPTION
**Pull request by staltz**

### CODE REVIEW

**Comment by staltz** [comment]
@TylorS can you review?

**Comment by TylorS** [comment]
Checking it now

**Comment by TylorS** [comment]
This looks good to merge :+1:

**Comment by TylorS** [comment]
I don't know which browsers you normally expect to fail on the travis-ci build though
Looks like a lot of them failed. Is that what you were referring to in #87?

**Comment by Frikki** [comment]
The failing tests on CI could very well be related to RxJS asynchronous behavior and timing with the assertions made. We also see test passing on one machine but not another, which seem to be related to speed of the machines. I suspect this to be the issue here, as well.

File: src/render-dom.js [hunk 1]
```diff
@@ -117,22 +117,23 @@ function makeEventsSelector(rootEl$, namespace) {
...
         const topSelector = namespace.join(``)
+        const roof = rootEl.parentElement
         return fromEvent(rootEl, eventName, useCapture).filter(ev => {
-          if (matchesSelector(ev.target, descendantSelector) ||
-              matchesSelector(ev.target, topSelector))
-          {
-            return isStrictlyInRootScope(ev.target)
+          for (let el = ev.target; el && el !== roof; el = el.parentElement) {
+            if (!isStrictlyInRootScope(el)) {
+              return false
```
Comment (laszlokorte) [inline src/render-dom.js]: I just tried running my app with the fix-event-delegation branch. Everything works fine except that my use case which is described in #80 is not working anymore. 
The problem seems to be the strictlyInRootScope check. But currently I am confused and not even sure why it was working before.
I am going to create a simple reproducible test case.

Comment (laszlokorte) [inline src/render-dom.js]: actually you can reuse the same use case i created for #80:
https://gist.github.com/laszlokorte/e2043a2544184d9aa928
it is not working anymore. Neither inner nor outer component get any of the events.

**Comment by TylorS** [comment]
There is a test for that though. Does this test do something differently?
https://github.com/cyclejs/cycle-dom/blob/master/test/browser/isolation.js#L232

**Comment by laszlokorte** [comment]
Oh, I think maybe it it not #80 that got broken but #82
The element for which I want to get the events is the root element of the component.

The test case for it (https://github.com/cyclejs/cycle-dom/blob/master/test/browser/isolation.js#L301) does only cover the *selection* of the element but not receiving it's events.

**Comment by laszlokorte** [comment]
There seem to be two (independent?) issues:

1. #82 is not working for the event delegation
2. If I have an hierarchy of isolated components:
Outer -> Middle -> Inner
And the Outer component is interested in click events.
I click on an element which is owned by Inner (but is inside an Element that is owned by Outer)
With classical event binding the Outer component would get the event because the StrictlyInRootScope scope was only done when adding the listener not when receiving the event. It did not matter if `event.target` is StrictlyInRootScope only the element the listener was added to had to satisfy.

Now with delegation `event.target` is checked for StrictlyInRootScope that check fails because the target could be inside a an isolated component.
I would expect that the Outer component still get's the event even if event.target is not strictly inside it's scope.
This case is not covered by https://github.com/cyclejs/cycle-dom/blob/master/test/browser/isolation.js#L232 because in the test case stopPropagation is used and it is not even expected that the frame(Outer) receives the monalisa(Inner) click.

**Comment by TylorS** [comment]
Thanks for testing it out @laszlokorte 
We'll get working on some more fixes :)

**Comment by staltz** [comment]
Thanks for pointing that out. We'll get all the appropriate tests and and make sure people will be happy before releasing v9.0.0.

**Comment by staltz** [comment]
@laszlokorte 
> I would expect that the Outer component still get's the event even if event.target is not strictly inside it's scope.

Sounds like parent-child isolation is simply just working. What you expect seems like parent-child isolation *with exceptions*. Can I suggest instead either:

- Don't isolate the child, and the parent is then able to see everything from the child.
- Or export a `click$` sink from the child, so the parent can use that.

**Comment by staltz** [comment]
I'm thinking/considering an option to make parent-child isolation opt-out.

```js
DOM.select('.foo') // cannot select .foo inside child
DOM.select('.foo', {childIsolation: false}) // CAN select .foo inside child
```

```js
DOM.select('.foo').events('click') // cannot capture clicks on .foo inside child
DOM.select('.foo').events('click', {childIsolation: false}) // CAN capture clicks on .foo inside child
```

What do people think about this? I can see how sibling isolation is almost always desired, and parent-child isolation is also by default preferred, but in some cases like Laszlo's, parent-child isolation is not desired.

**Comment by staltz** [comment]
One more commit pushed for review, @TylorS.

**Comment by TylorS** [comment]
PR look good @staltz 

Would `childIsolation: false` just skip the `isStrictlyInRootScope()` check?

**Comment by staltz** [comment]
> Would `childIsolation: false` just skip the `isStrictlyInRootScope()` check?

Yes (maybe there is something else too, but in principle yes).

**Comment by staltz** [comment]
What does @laszlokorte think of the proposed API?

**Comment by laszlokorte** [comment]
@staltz I am not sure you understood the issue I described correctly. I do not want opt out of isolation. The behavior of the current rootScope check simply does not match the event bubbling that happens in classical dom event binding
isolation should not stop event's from bubbling up. In my case I am talking about a mousewheel event. I have my canvas and the elements inside the canvas. I want the canvas to handle the wheel event (for zooming) currently (with delegation) zooming works only if the cursor is over a blank spot on the canvas but not during the mouse if over a element. That is because the strictlyInRoot check prevents the wheel event (which has target set to the child element) from being handled by the canvas.

When binding the listeners in a classical way the isolation would indeed prevent the canvas from querying DOM.select('.someIsolatedChildElement').events('wheel') but it would still be possible (and thats what I want) for the canvas to query DOM.select('.canvasRoot').events('wheel') because the wheel events bubbles up to the canvasRoot (unless propagation has been stopped by a child element's listener). The isolation should not interfere with that plain dom bubbling.

**Comment by laszlokorte** [comment]
I think that's what I am proposing:
```diff
 for (var el = ev.target; el && el !== roof; el = el.parentElement) {
-  if (!isStrictlyInRootScope(el)) {
-    return false;
-  }
-  if (matchesSelector(el, descendantSelector) || matchesSelector(el, topSelector)) {
-    Object.defineProperty(ev, "currentTarget", { value: el });
-    return true;
+  if (isStrictlyInRootScope(el)) {
+    if (matchesSelector(el, descendantSelector) || matchesSelector(el, topSelector)) {
+      Object.defineProperty(ev, "currentTarget", { value: el, configurable: true });
+      return true;
+    }
   }
 } 

```

**Comment by staltz** [comment]
@laszlokorte I can see clearly what you want. But Isolation is really a strong guarantee. Maybe it wasn't stated obviously enough what isolation is: a complete inability of detecting the presence on the DOM of another concurrent dataflow component. There are two types: sibling isolation (almost always desired), and parent-child isolation. The latter, parent-child, also silents event bubbling. That's because if we actually want isolation (read: complete ignorance of the DOM existence of the child dataflow component), then the parent should not be able to detect any DOM events coming from the child, even through bubbling.

> I want the canvas to handle the wheel event (for zooming) currently (with delegation) zooming works only if the cursor is over a blank spot on the canvas but not during the mouse if over a element.

Options are:
- The child captures wheel event and exports it as a sink back to the parent. Parent-child isolation turned on.
- Opt-out of parent-child isolation with `.events('mousewheel', {childIsolation: false'})`. Or do you actually want both parent-child isolation + bubbling from child to parent?

**Comment by laszlokorte** [comment]
> Or do you actually want both parent-child isolation + bubbling from child to parent?
yes. That is the behavior of 8.0.0 and that's what I want.

Image the following:

```
parentComponent
  ul
    li
     someIsolatedChild1
    li
     someIsolatedChild2
```
the parent want to listen for click events on the list items.
Maybe the isolatedChild has some own click listeners. Of course the parentComponent should NOT be able to listener for click events on *specific* child elements inside the isolated component.
But the parent should be able to listen for clicks on the list items (which it owns) and be notified about all the clicks that happen (except the ones that have explicitly been handled by isolatedChild and *stopped from propagation* (via stopPropagation).
The current event delegation does not notify the parentComponent about any clicks that happen unless they append outside the isolatedChild (which can only be the case if there is padding/margin around isolatedChild causing whitespace).

Why do you think the events should be withheld? They happen inside the bounding box of parent so parent should know about them.

**Comment by staltz** [comment]
Hmm, makes sense, but I'll have to think about it. All previous tests should pass too. We need to find a clear definition of what parent-child isolation means.

**Comment by laszlokorte** [comment]
I propose the following change and test case:
https://gist.github.com/laszlokorte/3ad1efbbd2a4579f4907

All other tests still pass.

**Comment by laszlokorte** [comment]
PS:
> Maybe it wasn't stated obviously enough what isolation is: a complete inability of detecting the presence on the DOM of another concurrent dataflow component.
In 8.0.0 the isolation was only the inability to `select` a DOM node of another dataflow component.

The whole issue I am talking about was just introduced with this delegation implementation and is not an issue of isolation by itself.

**Comment by staltz** [comment]
@laszlokorte you are right, thanks for the gist. I think I'm now confident that Cycle DOM isolation concerns `select()` only, not `events()`.

**Comment by staltz** [comment]
I pushed some commits. Now is time for some code reviews before this is in shape to be shipped.

**Comment by TylorS** [comment]
All of it looks good to merge to me. 

FWIW the 1500 slider example renders at 20 FPS (down from 35 FPS) with the updates to simulate event bubbling. Likely has to do with all the extra traversing involved.

**Comment by laszlokorte** [comment]
Now everything is working again in my application.

**Comment by staltz** [comment]
> FWIW the 1500 slider example renders at 20 FPS (down from 35 FPS) with the updates to simulate event bubbling. Likely has to do with all the extra traversing involved.

Was it this?
```diff
-return false
+continue
```

**Comment by TylorS** [comment]
I think is was
```diff
-break
+continue
```
but yes.

**Comment by laszlokorte** [comment]
if those few extra iterations caused by the `continue` are decreasing your performance you should be able to prevent them by actually calling `stopPropagation` on the events you are listening for.

**Comment by TylorS** [comment]
Yeah, I'm just reporting the changes when using the default settings with the 1500 slider example that Keyamoon gave us. It's where all of the event delegation started from. It used to be almost impossibly slow.

**Comment by laszlokorte** [comment]
Does stopping the propagation manually resolve the the regression?

**Comment by TylorS** [comment]
It might actually be an improvement :+1:

**Comment by TylorS** [comment]
Seeing about ~35-40 with just adding `.do(ev => ev.stopPropagation())`

**Comment by staltz** [comment]
:abcd: good :goat:

**Comment by laszlokorte** [comment]
:+1:

**Comment by staltz** [comment]
Merged.

**Pull request merged**